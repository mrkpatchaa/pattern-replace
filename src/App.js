import React from 'react';

import replace from './main';

import './App.scss';

class App extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      option: 'uuidv1',
      pattern: 'uuid',
      patternError: '',
      replacementError: '',
    };
  }

  handleOptionChange = event => {
    this.setState({option: event.target.value});
  };

  handlePatternChange = event => {
    this.setState({pattern: event.target.value});
  };

  handleReplace = () => {
    this.setState({patternError: '', replacementError: ''}, () => {
      if (this.state.pattern === 'custom' && !this.customPattern.value) {
        this.setState({patternError: 'Pattern is mandatory'});
        return;
      }
      if (this.state.option === 'custom' && !this.customReplacement.value) {
        this.setState({replacementError: 'Replacement text is mandatory'});
        return;
      }
      this.outputText.value = '';
      this.outputText.value = replace(
        this.sourceText.value,
        this.state.pattern === 'custom'
          ? this.customPattern.value
          : this.state.pattern,
        this.state.option === 'custom'
          ? this.customReplacement.value
          : this.state.option,
      );
    });
  };

  render() {
    return (
      <div className="row component-app h-100 pt-2 pb-2">
        <div className="col-sm">
          <div className="border border-primary rounded bg-white h-100">
            <form className="h-100">
              <div className="form-group h-100">
                <textarea
                  className="code form-control h-100"
                  ref={text => (this.sourceText = text)}
                />
              </div>
            </form>
          </div>
        </div>
        <div className="col-sm-3">
          <div className="border border-primary rounded h-100">
            <div className="align-middle">
              <form>
                <div className="form-group">
                  <label
                    htmlFor="replaceOption"
                    className="text-white text-uppercase font-weight-bold">
                    Pattern
                  </label>
                  <select
                    value={this.state.pattern}
                    className="form-control"
                    id="pattern"
                    onChange={this.handlePatternChange}>
                    <option value={'uuid'}>UUID</option>
                    <option value={'custom'}>Custom</option>
                  </select>
                </div>
                {this.state.pattern === 'custom' && (
                  <div className="form-group">
                    <input
                      type="text"
                      className="form-control"
                      placeholder="Custom pattern"
                      id="customPattern"
                      ref={input => (this.customPattern = input)}
                    />
                    {this.state.patternError.length > 0 && (
                      <div className="alert alert-danger p-1 mt-1" role="alert">
                        {this.state.patternError}
                      </div>
                    )}
                  </div>
                )}
                <div className="form-group">
                  <label
                    htmlFor="replaceOption"
                    className="text-white text-uppercase font-weight-bold">
                    Replace with
                  </label>
                  <select
                    value={this.state.value}
                    className="form-control"
                    id="replaceOption"
                    onChange={this.handleOptionChange}>
                    <option value={'uuidv1'}>UUID (V1)</option>
                    <option value={'uuidv4'}>UUID (V4)</option>
                    <option value={'custom'}>Custom text</option>
                  </select>
                </div>
                {this.state.option === 'custom' && (
                  <div className="form-group">
                    <input
                      type="text"
                      className="form-control"
                      placeholder="Replacement text"
                      id="customReplacement"
                      ref={input => (this.customReplacement = input)}
                    />
                    {this.state.replacementError.length > 0 && (
                      <div className="alert alert-danger p-1 mt-1" role="alert">
                        {this.state.replacementError}
                      </div>
                    )}
                  </div>
                )}
              </form>
              <button
                onClick={this.handleReplace}
                type="button"
                id="replaceButton"
                className="btn btn-outline-light btn-lg btn-block">
                Replace
              </button>
            </div>
          </div>
        </div>
        <div className="col-sm">
          <div className="border border-primary rounded bg-white h-100">
            <form className="h-100">
              <div className="form-group h-100">
                <textarea
                  className="code form-control h-100"
                  id="outputText"
                  ref={text => (this.outputText = text)}
                />
              </div>
            </form>
          </div>
        </div>
      </div>
    );
  }
}

export default App;
