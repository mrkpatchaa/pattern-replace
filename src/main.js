const uuidv1 = require('uuid/v1');
const uuidv4 = require('uuid/v4');

export const UUID_PATTERN =
  '[a-f0-9]{8}-?[a-f0-9]{4}-?[1-5][a-f0-9]{3}-?[89ab][a-f0-9]{3}-?[a-f0-9]{12}';

export const getArray = (size = 1) => Array.from(Array(size).keys());

export const isFunction = replacementOption =>
  replacementOption.startsWith('uuid');

export const getReplacementValue = option => {
  switch (option) {
    case 'uuidv1':
      return uuidv1();
    case 'uuidv4':
      return uuidv4();
    default:
      throw new Error('Unkown replacement function');
  }
};

export const getPattern = pattern => {
  switch (pattern) {
    case 'uuid':
      return UUID_PATTERN;
    default:
      return pattern;
  }
};

const replace = (content, pattern, replacement = 'uuidv1') => {
  if (!content) {
    return 'Empty content !';
  }

  if (!isFunction(replacement)) {
    return content.replace(new RegExp(pattern, 'gi'), replacement);
  }

  const match = content.match(new RegExp(getPattern(pattern), 'gi'));
  if (!match) {
    return content;
  }
  const uniques = [...new Set(match)];

  const replacements = getArray(uniques.length).map(() =>
    getReplacementValue(replacement),
  );

  let newContent = content;
  uniques.forEach((item, idx) => {
    newContent = newContent.replace(new RegExp(item, 'gi'), replacements[idx]);
  });
  return newContent;
};

export default replace;
