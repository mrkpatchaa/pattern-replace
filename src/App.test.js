import React from 'react';
import renderer from 'react-test-renderer';
import {shallow, mount} from 'enzyme';

import App from './App';
import * as main from './main';

test('Renders the application', () => {
  const component = renderer.create(<App />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('State changes when an option is selected', () => {
  const component = shallow(<App />);

  const event = {target: {value: 'uuidv4'}};
  component.find('#replaceOption').simulate('change', event);

  expect(component.state().option).toEqual('uuidv4');
});

test('State changes when a pattern is selected', () => {
  const component = shallow(<App />);

  const event = {target: {value: 'custom'}};
  component.find('#pattern').simulate('change', event);

  expect(component.state().pattern).toEqual('custom');
  expect(component.find('#customPattern').length).toEqual(1);
});

test('Show message error when custom pattern field is empty', () => {
  const component = mount(<App />);

  const event = {target: {value: 'custom'}};
  component.find('#pattern').simulate('change', event);
  component.find('#replaceButton').simulate('click');

  expect(component.state().patternError).toEqual('Pattern is mandatory');
  expect(component.find('.alert').text()).toEqual('Pattern is mandatory');
});

test('Show message error when custom pattern field is empty', () => {
  const component = mount(<App />);

  const event = {target: {value: 'custom'}};
  component.find('#replaceOption').simulate('change', event);
  component.find('#replaceButton').simulate('click');

  expect(component.state().replacementError).toEqual(
    'Replacement text is mandatory',
  );
  expect(component.find('.alert').text()).toEqual(
    'Replacement text is mandatory',
  );
});

test('Shows content replaced', () => {
  const component = mount(<App />);
  main.default = jest.fn().mockReturnValue('Replaced content');

  component.find('#replaceButton').simulate('click');

  expect(component.instance().outputText.value).toEqual('Replaced content');
});
