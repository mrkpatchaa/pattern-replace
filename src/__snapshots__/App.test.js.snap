// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the application 1`] = `
<div
  className="row component-app h-100 pt-2 pb-2"
>
  <div
    className="col-sm"
  >
    <div
      className="border border-primary rounded bg-white h-100"
    >
      <form
        className="h-100"
      >
        <div
          className="form-group h-100"
        >
          <textarea
            className="code form-control h-100"
          />
        </div>
      </form>
    </div>
  </div>
  <div
    className="col-sm-3"
  >
    <div
      className="border border-primary rounded h-100"
    >
      <div
        className="align-middle"
      >
        <form>
          <div
            className="form-group"
          >
            <label
              className="text-white text-uppercase font-weight-bold"
              htmlFor="replaceOption"
            >
              Pattern
            </label>
            <select
              className="form-control"
              id="pattern"
              onChange={[Function]}
              value="uuid"
            >
              <option
                value="uuid"
              >
                UUID
              </option>
              <option
                value="custom"
              >
                Custom
              </option>
            </select>
          </div>
          <div
            className="form-group"
          >
            <label
              className="text-white text-uppercase font-weight-bold"
              htmlFor="replaceOption"
            >
              Replace with
            </label>
            <select
              className="form-control"
              id="replaceOption"
              onChange={[Function]}
              value={undefined}
            >
              <option
                value="uuidv1"
              >
                UUID (V1)
              </option>
              <option
                value="uuidv4"
              >
                UUID (V4)
              </option>
              <option
                value="custom"
              >
                Custom text
              </option>
            </select>
          </div>
        </form>
        <button
          className="btn btn-outline-light btn-lg btn-block"
          id="replaceButton"
          onClick={[Function]}
          type="button"
        >
          Replace
        </button>
      </div>
    </div>
  </div>
  <div
    className="col-sm"
  >
    <div
      className="border border-primary rounded bg-white h-100"
    >
      <form
        className="h-100"
      >
        <div
          className="form-group h-100"
        >
          <textarea
            className="code form-control h-100"
            id="outputText"
          />
        </div>
      </form>
    </div>
  </div>
</div>
`;
